soc/nordic/nrf54h/power.c

```diff
@@ -0,0 +1,159 @@
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+}
+
+void nrf_poweroff(void)
+{
+	nrf_resetinfo_resetreas_local_set(NRF_RESETINFO, 0);
+	nrf_resetinfo_restore_valid_set(NRF_RESETINFO, false);
+
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+
+	/* TODO: Move it around k_cpu_idle() implementation. */
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+
+	suspend_common();
+
+	nrf_lrcconf_task_trigger(NRF_LRCCONF010, NRF_LRCCONF_TASK_SYSTEMOFFREADY);
+
+	__set_BASEPRI(0);
+	__ISB();
+	__DSB();
+	__WFI();
+
+	CODE_UNREACHABLE;
+}
+
+__weak void z_pm_soc_resume(void)
+{
+}
+
+/* Resume domain after local suspend to RAM. */
+void z_pm_sys_resume(void)
+{
+	if (IS_ENABLED(CONFIG_ICACHE)) {
+		/* Power up and re-enable ICACHE */
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+							RAMBLOCK_CONTROL_BIT_ICACHE, true);
+		sys_cache_instr_enable();
+	}
+
+	if (IS_ENABLED(CONFIG_DCACHE)) {
+		/* Power up and re-enable DCACHE */
+		nrf_memconf_ramblock_control_enable_set(NRF_MEMCONF, RAMBLOCK_POWER_ID,
+							RAMBLOCK_CONTROL_BIT_DCACHE, true);
+		sys_cache_data_enable();
+	}
+
+	/* Re-enable domain retention. */
+	nrf_lrcconf_retain_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, true);
+
+	/* TODO: Move it around k_cpu_idle() implementation. */
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN,
+			!IS_ENABLED(CONFIG_SOC_NRF54H20_CPURAD));
+
+	z_pm_soc_resume();
+}
+
+/* Function called during local domain suspend to RAM. */
+int z_pm_sys_suspend_to_ram(void)
+{
+	/* Set intormation which is used on domain wakeup to determine if resume from RAM shall
+	 * be performed.
+	 */
+	nrf_resetinfo_resetreas_local_set(NRF_RESETINFO,
+					  NRF_RESETINFO_RESETREAS_LOCAL_UNRETAINED_MASK);
+	nrf_resetinfo_restore_valid_set(NRF_RESETINFO, true);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_DOMAIN_0, false);
+	nrf_lrcconf_poweron_force_set(NRF_LRCCONF010, NRF_LRCCONF_POWER_MAIN, false);
+
+	suspend_common();
+
+	__set_BASEPRI(0);
+	__ISB();
+	__DSB();
+	__WFI();
+	/*
+	 * We might reach this point is k_cpu_idle returns (there is a pre sleep hook that
+	 * can abort sleeping.
+	 */
+	return -EBUSY;
+}
+
+__weak bool z_pm_do_suspend(void)
+{
+	return true;
+}
+
+void do_suspend(void)
+{
+	if (!z_pm_do_suspend()) {
+		return;
+	}
+
+	/*
+	 * Save the CPU context (including the return address),set the SRAM
+	 * marker and power off the system.
+	 */
+	(void)soc_s2ram_suspend(z_pm_sys_suspend_to_ram);
+
+	/*
+	 * On resuming or error we return exactly *HERE*
+	 */
+
+	z_pm_sys_resume();
+}
+
+void pm_state_set(enum pm_state state, uint8_t substate_id)
+{
+	if (state != PM_STATE_SUSPEND_TO_RAM) {
+		return;
+	}
+#if IS_ENABLED(CONFIG_PM_S2RAM)
+	do_suspend();
+#endif
```

Comment:
If you are not doing anything you should call `k_cpu_idle()` otherwise the idle thread will loop on it until the next scheduled time is lower than the suspend-to-ram residency time.

What does not make much sense is having PM enabled if you don't handle any power state, you could simply disable PM when `CONFIG_PM_S2RAM` is disabled.

Note that device power management does not require `CONFIG_PM`

Reply:
> What does not make much sense is having PM enabled if you don't handle any power state, you could simply disable PM when `CONFIG_PM_S2RAM` is disabled.
But [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/pm/Kconfig#L34) is reversed dependence.